idaes/core/base/property_set.py

```diff
@@ -0,0 +1,2099 @@
+        """
+        Set method attribute of property.
+
+        Args:
+            meth: reference to method required to construct this property
+
+        Returns:
+            None
+        """
+        # TODO: Validate that meth is callable?
+        self._method = meth
+
+    def set_supported(self, supported=True):
+        """
+        Set supported attribute of property
+
+        Args:
+            supported: bool indicating whether package supports this property
+
+        Returns:
+            None
+        """
+        # TODO: Validate that supported is bool
+        self._supported = supported
+
+    def set_required(self, required=True):
+        """
+        Set required attribute of property
+
+        Args:
+            required: bool indicating whether package requires this property be defined by
+            another property package
+
+        Returns:
+            None
+        """
+        # TODO: Validate that required is bool
+        self._required = required
+
+    def update_property(self, method=None, required=None, supported=None):
+        """
+        Update attributes of this property.
+
+        Args:
+            method: method name (str) to be assigned to construct property
+            required : bool indicating whether this property package requires this property to be
+                defined by another package
+            supported : bool indicating whether this property package supports this property
+
+        Returns:
+            None
+
+        Note that if not provided a value, 'supported' is assumed to be True.
+        """
+        if method is not None:
+            self.set_method(method)
+        if required is not None:
+            self.set_required(required)
+        if supported is not None:
+            self.set_supported(supported)
+        else:
+            # Assume supported if not explicitly stated
+            # TODO: Reconsider in the future, for now do this for backwards compatibility
+            self.set_supported(True)
+
+
+class PropertySetBase(object):
+    """
+    Base class for defining property sets.
+
+    This defines the common methods expected of all PropertySets.
+    """
+
+    def __init__(self, parent):
+        self._parent_block = parent
+
+    def __getitem__(self, key):
+        try:
+            return getattr(self, key.upper())
+        except AttributeError:
+            try:
+                return getattr(self, key)
+            except AttributeError:
+                raise KeyError(f"Property {key} is not defined in this PropertySet.")
+
+    def __iter__(self):
+        for a in dir(self):
+            if not a.startswith("_"):
+                # Ignore anything starting with "_" as they will show up twice otherwise
+                # This will also skip private attributes of object class
+                aobj = getattr(self, a)
+                if isinstance(aobj, PropertyMetadata):
+                    yield aobj
+
+    def define_property(
```

Comment:
The docs say name is required, yet the code seems to silently use the default value of None. Suggest taking `=None` off the name argument to enforce this.